v21.1 Домашнее задание спринт №21

KravtsovSO64 opened the pull request.

Commit:
v21.1 Домашнее задание спринт №21

IliaAnK commented on app/src/main/java/com/practicum/playlistmaker/data/db/dao/FavouriteTrackDao.kt:
```diff
@@ -0,0 +1,25 @@
+    suspend fun deleteTrack(track: FavouriteTrackEntity)
+
+    @Query("SELECT * FROM  favourite_track_table ORDER BY timestamp DESC")
+    suspend fun getListFavouriteTracks(): List<FavouriteTrackEntity>
```
Рекомендация:
`Room` может возвращать `Flow`
Получится:
`fun getListFavouriteTracks(): Flow<List<FavouriteTrackEntity>>`

KravtsovSO64 replied:
а в чем различие? Честно говоря спринт за спринтом идут, мало было времени разобраться. Поэтому применил самый простой на мой взгляд прием "Указать модификатор функции suspend". С потоком Flow<List<Track>>, можно было бы данные в процессе фильтровать и модифицировать, я так понимаю различие между двумя подходами в этом?!

IliaAnK replied:
Разница в том, что так флоу формируется на стороне библиотеки Room. При этом поток получает все изменения, которые происходят в БД, что постоянно поддерживает актуальность данных

IliaAnK commented on app/src/main/java/com/practicum/playlistmaker/ui/media/FragmentFavoriteTracks.kt:
```diff
@@ -0,0 +1,129 @@
+
+        //Recycler and Adapter
+        val recyclerFavouriteTrack = binding.favouriteTrackRecycler
+        recyclerFavouriteTrack.layoutManager = LinearLayoutManager(requireContext(), LinearLayoutManager.VERTICAL, false)
```
Рекомендация:
`layoutManager` можно задать в вёрстке, через одноимённый атрибут

KravtsovSO64 replied:
Так привычней уже. Учту в дальнейшем!

IliaAnK commented on app/src/main/java/com/practicum/playlistmaker/ui/media/uicomponents/FavouriteTracksAdapter.kt:
```diff
@@ -0,0 +1,31 @@
+    override fun onBindViewHolder(holder: FavouriteTracksViewHolder, position: Int) {
+        holder.bind(favouriteListAdapter[position])
+        holder.itemView.setOnClickListener {
+            listener.onItemClick(position)
```
Рекомендация:
Вместо позиции лучше сразу передавать модель данных, тогда не нужно будет "размазывать" логику извлечения нужной модели по разным классам. В таком случае, метод `getList` можно будет удалить
Пример:
`favouriteListAdapter[position]`

KravtsovSO64 replied:
Если будет время обязательно сделаю рефакторинг. Уже опаздываю со следующим спринтом. Увы

KravtsovSO64 replied:
Исправил. В этой гонке за спринтами. Порой мухи не заметишь перед носом. Стало гораздо лучше.

KravtsovSO64 replied:
В общем за двое суток, что я только не перепробовал с данным кейсом.
Кейс: перейти на экран "Поиск" -> выполнить поиск -> нажать на любой найденный трек -> открылся экран "Аудиоплеер" -> добавить трек в избранное и нажать кнопку "Назад" -> произошёл возврат на экран "Поиск" -> снова открыть тот же трек.
Единственное, что работает это переделать экран плеера под фрагмент. Передавать через Bundel Track. И я даже научился его возвращать с помощью setFragmentResultListener. Но потом уперся в огромную стену, с хранением списка поискового запроса в viewModel и его обновлении. Почему каждый раз при переходе на экран плеера, viewModel экрана поиска, видимо тоже уничтожалась вместе с экраном. Ведь при возвращении на экран поиска с плеера, список из адаптера был пуст, список при получении из viewModel сохраненные там до перехода тоже был пуст. 
В итоге получилось как получилось.

KravtsovSO64 replied:
Список адаптера понятно, почему пустой!

IliaAnK replied:
Кажется, здесь нужно просто сделать запрос в БД с id открываемого трека, чтобы проверить его наличие в избранном.

Commit:
v21.2 Домашнее задание спринт №21

IliaAnK commented on app/src/main/java/com/practicum/playlistmaker/domain/model/Track.kt:
```diff
@@ -1,6 +1,7 @@
 package com.practicum.playlistmaker.domain.model
 
-import java.io.Serializable
+import android.os.Parcel
```
Обязательно:
Согласно требованиям 15 и 16 спринтов, в `domain` слое не должно быть зависимостей на `Android SDK`

KravtsovSO64 replied:
Исправил

Commit:
v21.2 Домашнее задание спринт №21

Commit:
v21.2 Домашнее задание спринт №21

Commit:
v21.2 Домашнее задание спринт №21

Commit:
v21.2 Домашнее задание спринт №21

Commit:
v21.3 Домашнее задание спринт №21
-Изменил архитектуру приложения на  Single Activity
-Создал общий shareViewModel, для того чтобы хранить список адаптера при переходе на экран плеера.
-Так же добавил функции обновления треков в поисковой выдаче. Теперь корректно отображается в плеере избранное.
-Из Domain убрал завивисмость от SDK
-Тестировал на Motorola moto g20.

Commit:
v21.3 Домашнее задание спринт №21

Commit:
v21.3 Домашнее задание спринт №21

Commit:
v21.3 Домашнее задание спринт №21

Commit:
v21.3 Домашнее задание спринт №21
Test

Commit:
v21.3 Домашнее задание спринт №21
